noir-projects/noir-contracts/contracts/key_registry_contract/src/main.nr

```diff
@@ -0,0 +1,144 @@
+contract KeyRegistry {
+  use dep::aztec::{
+    avm::hash::poseidon,
+    context::gas::GasOpts,
+    state_vars::{
+      SharedMutable,
+      Map
+    }, 
+    protocol_types::{
+      abis::function_selector::FunctionSelector,
+      contract_class_id::ContractClassId,
+      address::{
+        AztecAddress,
+        EthAddress,
+        PublicKeysHash,
+        PartialAddress,
+      },
+      constants::{
+        GENERATOR_INDEX__CONTRACT_ADDRESS,
+        GENERATOR_INDEX__PARTIAL_ADDRESS
+      },
+    },
+  };
+
+  global KEY_ROTATION_DELAY = 5;
+
+  #[aztec(storage)]
+  struct Storage {
+    nullifier_public_key_registry: Map<AztecAddress, SharedMutable<Field, KEY_ROTATION_DELAY>>,
+    incoming_public_key_registry: Map<AztecAddress, SharedMutable<Field, KEY_ROTATION_DELAY>>,
+    outgoing_public_key_registry: Map<AztecAddress, SharedMutable<Field, KEY_ROTATION_DELAY>>,
+    tagging_public_key_registry: Map<AztecAddress, SharedMutable<Field, KEY_ROTATION_DELAY>>,
+  }
+
+  #[aztec(public)]
+  fn register(
+    nullifier_public_key: Field,
+    incoming_public_key: Field,
+    outgoing_public_key: Field,
+    tagging_public_key: Field,
+  ) {
```

Comment:
You always want to do the check that the hash matches real account address so there should be only 1 register function.